### PR TITLE
add rcScopedTimer to fix timers not being stopped on all paths and to…

### PR DESCRIPTION
**Pull request by rafzi**
… prevent it in the future.

This fixes various cases where timers are not stopped.

Required for #102 

Notes:
- In rcBuildDistanceField behaviour is changed, because the timer runs until after rcFree, but it's more correct now.
- In rcBuildRegions and rcBuildContours some start/stop calls are left as is, because the code would become uglier with no benefit.

### CODE REVIEW

File: Recast/Source/RecastRegion.cpp [hunk 1]
```diff
@@ -1489,23 +1487,21 @@ bool rcBuildRegionsMonotone(rcContext* ctx, rcCompactHeightfield& chf,
...
-	ctx->startTimer(RC_TIMER_BUILD_REGIONS_FILTER);
+	{
+		rcScopedTimer timerFilter(ctx, RC_TIMER_BUILD_REGIONS_FILTER);
 
-	// Merge regions and filter out small regions.
-	rcIntArray overlaps;
-	chf.maxRegions = id;
-	if (!mergeAndFilterRegions(ctx, minRegionArea, mergeRegionArea, chf.maxRegions, chf, srcReg, overlaps))
-		return false;
+		// Merge regions and filter out small regions.
```
Comment (hymerman) [inline Recast/Source/RecastRegion.cpp]: Indentation

File: Recast/Source/RecastRegion.cpp [hunk 2]
```diff
@@ -1802,22 +1796,20 @@ bool rcBuildLayerRegions(rcContext* ctx, rcCompactHeightfield& chf,
...
-	rcIntArray overlaps;
-	chf.maxRegions = id;
-	if (!mergeAndFilterLayerRegions(ctx, minRegionArea, chf.maxRegions, chf, srcReg, overlaps))
-		return false;
-	
-	ctx->stopTimer(RC_TIMER_BUILD_REGIONS_FILTER);
+	{
+		rcScopedTimer timerFilter(ctx, RC_TIMER_BUILD_REGIONS_FILTER);
+
+		// Merge monotone regions to layers and remove small regions.
```
Comment (hymerman) [inline Recast/Source/RecastRegion.cpp]: Indentation

**Comment by hymerman** [comment]
Looks good! Thanks @rafzi! Just minor formatting tweaks to fix then I'll merge in.

**Comment by grahamboree** [comment]
Awesome!  I really like this change.  It makes the code simpler and more idiomatic.  Looking over the areas @hymerman commented about indentation, they seemed fine to me?  Maybe I need more :coffee:?

**Comment by Janiels** [comment]
LGTM as well. :+1:

**Comment by hymerman** [comment]
Weird. They look fine in the conversation tab, but not in the 'files changed' or commits tabs - what do you guys see there?

Maybe there are mixed tabs and spaces and github is automatically converting them to tabs in some views and not in others?

**Comment by grahamboree** [comment]
That's possible.  I think github uses 8 space equivalent hard tabs (gross).  The context with the deleted lines can also make it look wrong.  I just cloned @rafzi's repo to verify, and things look fine there.  We can always fix this later if it's screwed up after merging.

**Comment by hymerman** [comment]
Yuck. Shame on github! Ok, merging in now.

**Comment by grahamboree** [comment]
:+1:

**Pull request merged**